rs/nns/governance/src/governance.rs

```diff
@@ -3315,23 +3315,86 @@ impl Governance {
             let topic = topic.unwrap_or_default();
             let followees = followees.clone();
 
-            if followees.is_empty() {
-                new_followees.remove(&topic);
-            } else {
-                new_followees.insert(topic, Followees { followees });
-            }
+            topic_to_followees = self.modify_followees(
+                &neuron,
+                &topic_to_followees,
+                topic,
+                Followees { followees },
+            )?;
         }
 
         // Commit new_followees to the neuron.
         let now_seconds = self.env.now();
         self.with_neuron_mut(id, |neuron| {
-            neuron.followees = new_followees;
+            // TODO use the function
+            neuron.followees = topic_to_followees;
             neuron.refresh_voting_power(now_seconds);
         })?;
 
         Ok(())
     }
 
+    fn modify_followees(
+        &self,
+        neuron: &Neuron,
+        topic_to_followees: &HashMap<i32, Followees>,
+        topic: i32,
+        new_followees: Followees,
+    ) -> Result<HashMap<i32, Followees>, GovernanceError> {
+        let controller = neuron.controller();
+        let hotkeys: &Vec<PrincipalId> = neuron.hot_keys.as_ref();
+        let mut updated_followees = topic_to_followees.clone();
+        if new_followees.followees.is_empty() {
+            // If the new followees list is empty, remove the entry for the topic.
+            updated_followees.remove(&topic);
+        } else {
+            // Otherwise, update the entry with the new followees list.
+            if is_neuron_indexes_enabled() {
+                // A new can follow another neuron if:
+                // 1. the followee neuron is a public neuron
+                // 2. or if the followee neuron is a private neuron, they share a controller.
+                // If in the list of followees, there is any follow relationship
+                // the doesn't adhere to the aforementioned rules, return a GovernanceError.
+                let (mut invalid_followees, mut error_message) = (0_u32, String::new());
+
+                for follow in &new_followees.followees {
+                    if let Ok(follow_neuron) = self.with_neuron(follow, |neuron| neuron.clone()) {
+                        if follow_neuron.visibility() == Visibility::Public
+                            || follow_neuron.controller() == controller
+                            || hotkeys.contains(&follow_neuron.controller())
+                        {
+                            continue;
+                        } else {
+                            invalid_followees += 1;
+                            error_message.push_str(&format!(
+                                "{}: Not authorized to follow private neuron: {follow:?}\n\t\
```

Comment:
How about adding a comment to prevent unexpected message text changes? Something like:
```
// Note: These error messages are matched in the nns-dapp. In case of changes, please sync with the nns dev team.
```